model.py

```diff
@@ -239,3 +314,184 @@ def forward(self, x):
         x = F.relu(x) * mix
 
         return x
+
+
+class Separator(nn.Module):
+    """
+    Separator class to encapsulate all the stereo filtering
+    as a torch Module, to enable end-to-end learning.
+
+    Parameters
+    ----------
+    targets: dictionary of target models {target: model}
+        the spectrogram models to be used by the Separator. Each model
+        may for instance be loaded with the `model.load_models` function
+
+    niter: int
+         Number of EM steps for refining initial estimates in a
+         post-processing stage. Zeroed if only one target is estimated.
+         defaults to 1.
+
+    residual: str or None
+        adds an additional residual target with provided name, obtained by
+        subtracting the other estimated targets from the mixture, before any
+        potential EM post-processing.
+        Defaults to None
+
+    out: str or None
+        if provided, must be a string containing a valid expression for
+        a dictionary, with keys the output target names, and values
+        a list of targets that are used to build it. For instance:
+        '{"vocals":["vocals"], "accompaniment":["drums","bass","other"]}'
+
+    batch_size: {None | int}
+        The size of the batches (number of frames) on which to apply filtering
+        independently. This means assuming time varying stereo models and
+        localization of sources.
+        None means not batching but using the whole signal. It comes at the
+        price of a much larger memory usage.
+    """
+    def __init__(
+        self,
+        targets,
+        niter=1,
+        softmask=False,
+        residual=None,
+        out=None,
+        batch_size=None
+    ):
+        super(Separator, self).__init__()
+        if not utils._torchaudio_available():
+            raise Exception('The Separator class only works when torchaudio '
+                            'is available.')
+
+        # saving parameters
+        self.niter = niter
+        self.residual = residual
+        self.out = None if out is None else json.loads(out)
```

Comment:
whats the purpose of this? 

1. we shouldn't use the `json` module here
2. we might better fiddle with the output dict after performing separation

Reply:
it was the most elegant way I found to create custom outputs from target models, what would be your proposition ?

Reply:
put this outside of the model and expect a dict.

Reply:
ok why not